AI/VCAI/Pathfinding/AIPathfinder.cpp

```diff
@@ -12,30 +12,50 @@
 #include "AIPathfinderConfig.h"
 #include "../../../CCallback.h"
 
+std::vector<std::shared_ptr<AINodeStorage>> AIPathfinder::storagePool = std::vector<std::shared_ptr<AINodeStorage>>();
+std::map<HeroPtr, std::shared_ptr<AINodeStorage>> AIPathfinder::storageMap = std::map<HeroPtr, std::shared_ptr<AINodeStorage>>();
+
 AIPathfinder::AIPathfinder(CPlayerSpecificInfoCallback * cb)
 	:cb(cb)
 {
-	nodeStorage.reset(new AINodeStorage(cb->getMapSize()));
 }
 
 void AIPathfinder::clear()
 {
-	currentHero = HeroPtr();
+	storageMap.clear();
 }
 
 std::vector<AIPath> AIPathfinder::getPathInfo(HeroPtr hero, int3 tile)
 {
-	if(!currentHero || currentHero != hero)
+
+	std::shared_ptr<AINodeStorage> nodeStorage;
+
+	if(!vstd::contains(storageMap, hero))
 	{
 		logAi->debug("Recalculate paths for %s", hero->name);
 
+		if(storageMap.size() < storagePool.size())
+		{
+			nodeStorage = storagePool.at(storageMap.size());
+		}
+		else
+		{
+			nodeStorage = std::make_shared<AINodeStorage>(cb->getMapSize());
+			storagePool.push_back(nodeStorage);
+		}
+
+		storageMap[hero] = nodeStorage;
+
 		boost::unique_lock<boost::mutex> pathLock(nodeStorage->getMutex());
```

Comment:
Where is a lock for storagePool?

Reply:
Good question. I also asked it myself. Do you think having a separate lock for storagePool and map is wise? Or should I use a single lock for both pool and all paths?

Reply:
You should use separate lock if it is possible to release it earlier than nodeStorage one.

Reply:
Does not look like this is wise. Storage in storageMap should either be already calculated or not present at all